### ISSUE TITLE
Testing workflow?

### ISSUE THREAD

**Issue by mockdeep** [rails/webpacker#318]
We're trying to set up Webpacker in our Rails 4.2 project and we've more or less got it working, but we've got some unanswered questions. In particular, if we're trying to run Capybara feature tests locally, do we always need to run `RAILS_ENV=test ./bin/rails webpacker:compile` beforehand? Or is there some other workflow that makes sense for this? One option we're considering is doing that on CI, but when we're developing locally having the tests fall back to relying on the dev server.

**Comment by jackjennings** [comment]
The assets are being compiled in the test environment during controller and system tests: https://github.com/rails/webpacker/blob/master/lib/tasks/webpacker/compile.rake#L30-L31

It seems like it would also make sense that this would happen when bare `rails test` is run, given that this command includes that invocation as well.

My workaround at this point is to add the following to the `Rakefile`:

```ruby
Rake::Task["test"].enhance(["webpacker:compile_before_test"])
```

Does it make sense to include this by default?

**Comment by mockdeep** [comment]
Ah, I see. We're still on Rails 4.2, but I guess we could add that to our `Rakefile` as well. We were actually taking an approach where we require `webpack-dev-server` to be running for the tests when testing locally, which also allows us to use `rspec spec/features/whatever_spec.rb` and not needing to go via `rake`/`rails`.

**Comment by jackjennings** [comment]
I noticed that works too, but having to run a separate process to get tests working is a very non-ideal developer experience.

**Comment by jackjennings** [comment]
The above workaround doesn't work with rspec (frustratingly). Not sure what to do there, and not familiar enough with rspec to recommend an avenue to support it, unfortunately.

**Comment by mockdeep** [comment]
Being that we're more likely than not to have the dev server already running, it isn't too much a pain on our end, and there's the upshot that it will also speed up spec runs by not needing to compile assets on each run. But it would be nice to have something that's a little more transparent.

**Comment by p0wl** [comment]
maybe you can benefit from this issue: https://github.com/rails/webpacker/issues/59. Especially with the latest version of webpacker, you may not be able to use the running webpack-dev-server in your test environment, because webpacker will search for `manifest-test.json` in `RAILS_ENV=test` (see https://github.com/rails/webpacker/blob/master/lib/install/config/webpack/paths.yml#L30)

**Comment by gauravtiwari** [comment]
No that should work because the manifest is written to file regardless of what binstub you use 👍  Will probably setup a base test helper to address to this

**Comment by mockdeep** [comment]
We changed `paths.yml` not to override `manifest` in test mode, so all environments write to `manifest.json`. This works out so that when testing locally the tests just use the dev server endpoint. On CI we precompile and set `Capybara.server_port = 8080` so that the Rails server serves up the assets from the public directory.

**Comment by returnlytom** [comment]
thanks @mockdeep, removing the `manifest` rule and adding this worked:

```rb
RSpec.configure do |config|
  config.before(:each, selenium: true) do
    Capybara.server_port = 8080
  end
end
```

**Comment by gauravtiwari** [comment]
After #360 this seems like a non-issue. Feel free to open another ticket if something else pops up 👍

**Comment by mockdeep** [comment]
Awesome, thanks @javan, @gauravtiwari! This looks like it'll do the trick.

**Comment by gauravtiwari** [comment]
Document this in README/guide - #372

**Comment by mcmire** [comment]
While the documentation gets updated, can someone tell me what the conclusion here is in terms of using Webpacker in a test environment? Is it as simple as running `RAILS_ENV=test bin/webpack-dev-server` in a separate process? Or do you have to configure Webpacker to do something fancy?

**Comment by toadkicker** [comment]
I just fought with this rspec and CircleCI. It came down to simply adding `- run: bin/webpack` to circle.yml
